### ISSUE TITLE
Stop processing expired requests in history service

### ISSUE THREAD

**Issue by tamer-eldeeb** [uber/cadence#319]
There are many places in the history service code where we block (e.g. waiting on a lock, or backoff retries on persistence). We should stop processing the request if the context has expired instead of trying to execute it after it has timed out.

**Comment by samarabbas** [comment]
Duplicate of #560
